e2e/cypress/plugins/smtpTester.js

```diff
@@ -29,6 +33,86 @@ const initialize = (on, config) => {
       return lastEmail[email] || null;
     },
   });
+
+  on('task', {
+    async compareImages({ expectedImageFilepath, newImageContent, updateExpectedImage = false }) {
+      const inputBuffer = Buffer.from(newImageContent.data);
+      if (updateExpectedImage) {
+        fs.writeFileSync(expectedImageFilepath, inputBuffer);
+        return true;
+      }
+
+      const inputImage = await Jimp.read(inputBuffer);
+      const expectedImage = await Jimp.read(expectedImageFilepath);
+
+      const pixelDiff = diff(expectedImage, inputImage, 0.3);
+      return pixelDiff.percent <= 0.1;
+    },
+  });
+
+  on('task', {
+    async compareCSVs({ expectedCSVFilepath, newCSVContent, updateExpectedCSV = false }) {
+      const inputBuffer = Buffer.from(newCSVContent.data);
+      if (updateExpectedCSV) {
+        await fs.writeFileSync(expectedCSVFilepath, inputBuffer);
+        return true;
+      }
+
+      const inputCSV = toCSV(inputBuffer);
+      const expectedCSV = toCSV(fs.readFileSync(expectedCSVFilepath));
+
+      if (inputCSV.length !== expectedCSV.length) {
+        return false;
+      }
+
+      for (let i = 0; i < expectedCSV.length; i++) {
+        const line = expectedCSV[i];
+        for (let j = 0; j < line.length; j++) {
+          if (line[j] !== inputCSV[i][j]) {
+            return false;
+          }
+        }
+      }
+
+      return true;
+    },
+  });
+
+  on('task', {
+    async comparePDFs({ expectedPDFFilepath, newPDFContent, updateExpectedPDF = false }) {
+      const inputBuffer = Buffer.from(newPDFContent.data);
+      if (updateExpectedPDF) {
+        fs.writeFileSync(expectedPDFFilepath, inputBuffer);
+        return true;
+      }
+
+      const inputDoc = await pdf(inputBuffer);
+      const expectedDoc = await pdf(expectedPDFFilepath);
+
+      cleanUpPDFText(inputDoc);
+      cleanUpPDFText(expectedDoc);
+
+      return inputDoc.numpages === expectedDoc.numpages && inputDoc.text === expectedDoc.text;
+    },
+  });
+};
+
+const toCSV = (buffer) => {
+  return buffer
+    .toString()
+    .split('\n')
+    .map((e) => e.trim())
+    .map((e) => e.split(',').map((e) => e.trim()));
+};
+
+const replacer = (match, p1, p2, p3) => {
+  return `${p1} ${p3}`;
+};
+
+// clean up the "generated on" timestamp in PDF as it's too complicated to set it to a fixed date
```

Comment:
I'm having a hard time understanding this code (regex and callbacks 😅). This removes the timestamp at the end or the `Generated on Friday, Nov 29 2024` part?
We could be more specific in the func comment for the future.

Reply:
I updated the comment and the function to hopefully make it easier to understand.